### PR TITLE
Printers for flint's arb and acb modules.

### PR DESCRIPTION
**Pull request by bjodah**
#### Brief description of what is fixed or changed
This is enough to get some basic test passing.

#### Other comments
Added this as a draft for now, hope to get back to this with more extensive testing soon.

#### Release Notes

* printing
  * Added ArbPrinter for use with `flint.arb`
  * Added AcbPrinter for use with `flint.acb`
* utilities
  * lambdify can now use `flint.arb`/`flint.acb` as a backend for ball arithmetic for real/complex numbers.

### CODE REVIEW

File: sympy/utilities/lambdify.py [hunk 1]
```diff
@@ -88,6 +90,9 @@
     "betainc_regularized": "betainc",
 }
 
+ARB_TRANSLATIONS = {
```
Comment (bjodah) [inline sympy/utilities/lambdify.py:93]:
```suggestion
ARB_TRANSLATIONS: dict[str, str]  = {
```

File: sympy/utilities/tests/test_lambdify.py [hunk 2]
```diff
@@ -1689,3 +1690,18 @@ def test_23536_lambdify_cse_dummy():
...
+def test_lambdify_arb():
+    if not flint:
+        skip("flint not installed.")
+
+    f1 = lambdify([x, y], x**3-y**3, modules='arb')
+    a1 = f1(10**15, 10**15 - 1)
+    a1_lo = float(flint.arb.abs_lower(a1))
+    a1_hi = float(flint.arb.abs_upper(a1))
+    assert math.isclose(a1_lo, 2.8610732685849445e+30)
+    assert math.isclose(a1_hi, 3.160267082499145e+30)
```
Comment (oscarbenjamin) [inline sympy/utilities/tests/test_lambdify.py:1703]: I still haven't found a good way of testing an Arb: https://github.com/fredrik-johansson/python-flint/pull/39#issuecomment-1357580852

**Comment by oscarbenjamin** [comment]
Testing this I found:
```python
In [6]: lambdify(x, factorial(x), modules='arb')(1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-e35040089efc> in <module>
----> 1 lambdify(x, factorial(x), modules='arb')(1)

<lambdifygenerated-7> in _lambdifygenerated(x)
      1 def _lambdifygenerated(x):
----> 2     return fac(x)

TypeError: descriptor 'fac' for 'flint._flint.arb' objects doesn't apply to a 'int' object
```
Should that be expected? It seems that the input must be an actual Arb.

**Comment by oscarbenjamin** [comment]
I made a script to try and test this systematically with every function from `sympy.functions`. The final output is:
```
total: 21400
skipped: 16319
passed: 3929
failed: 1152
```


```python
from random import normalvariate
from sympy import *
from flint import arb

# Everything from sympy.functions.__all__

functions = [
    factorial, factorial2, rf, ff, binomial, RisingFactorial, FallingFactorial,
    subfactorial, carmichael, fibonacci, lucas,
    # XXX: motzkin not a symbolic function
    #motzkin,
    tribonacci, harmonic,
    bernoulli, bell, euler, catalan, genocchi, andre, partition, sqrt, root,
    Min, Max, Id, real_root, cbrt, Rem, re, im, sign, Abs, conjugate, arg,
    # XXX: Not sure any of these make sense for mpmath/arb:
    #polar_lift, periodic_argument, unbranched_argument, principal_branch,
    #transpose, adjoint, polarify, unpolarify,
    sin, cos, tan, sec, csc, cot,
    sinc, asin, acos, atan, asec, acsc, acot, atan2,
    # exp_polar,
    exp, ln, log,
    LambertW, sinh, cosh, tanh, coth, sech, csch, asinh, acosh, atanh, acoth,
    asech, acsch, floor, ceiling, frac,
    # Probably Piecewise should be tested, needs special consideration for Arb
    # Piecewise,
    # Non-symbolic functions
    # piecewise_fold, piecewise_exclusive,
    erf, erfc, erfi, erf2, erfinv, erfcinv, erf2inv, Ei,
    expint, E1, li, Li, Si, Ci, Shi, Chi, fresnels, fresnelc, gamma,
    lowergamma, uppergamma, polygamma, loggamma, digamma, trigamma, multigamma,
    dirichlet_eta, zeta, lerchphi, polylog, stieltjes, riemann_xi,
    # These possibly don't make sense for mpmath/arb
    #Eijk, LeviCivita, KroneckerDelta, SingularityFunction, DiracDelta,
    Heaviside,
    # These possibly don't make sense for mpmath/arb
    # bspline_basis, bspline_basis_set, interpolating_spline,
    besselj, bessely,
    besseli, besselk, hankel1, hankel2, jn, yn,
    # Non-symbolic
    # jn_zeros,
    hn1, hn2, airyai,
    airybi, airyaiprime, airybiprime, marcumq,
    # These have different signatures so need bespoke testing:
    # hyper, meijerg, appellf1,
    legendre, assoc_legendre, hermite, hermite_prob, chebyshevt, chebyshevu,
    # Non-symbolic
    #chebyshevu_root, chebyshevt_root,
    laguerre, assoc_laguerre, gegenbauer,
    jacobi,
    # Non-symbolic
    # jacobi_normalized,
    Ynm,
    # Non-symbolic
    # Ynm_c,
    Znm, elliptic_k, elliptic_f,
    elliptic_e, elliptic_pi, beta, betainc, betainc_regularized, mathieus,
    mathieuc, mathieusprime, mathieucprime
]

values = [0, 1, 2]
values += [normalvariate(0, 1) for _ in range(10)]
values = [v for v in values if v >= 0]

x, y, z = symbols('x, y, z')

def pairs1(v):
    return [(v, vp) for vp in values]

def pairs2(v):
    return [(vp, v) for vp in values]

def allpairs():
    return [(v1,v2) for v1 in values for v2 in values]

def alltriples():
    return [(v1,v2,v3) for v1 in values for v2 in values for v3 in values]

failures = {
    # TypeError: unsupported operand type(s) for %
    factorial2: values,
    frac: values,
    # ValueError: gamma function pole (probably due to rewrite)
    rf: pairs1(0),
    ff: pairs1(0) + [(1,2)],
    binomial: [(0,1), (0,2), (1,2)],
    # e.g. NameError: name 'uppergamma' is not defined
    # TypeError: f() takes 1 positional argument but 2 were given
    subfactorial: values,
    carmichael: values + allpairs(),
    fibonacci: values + allpairs(),
    lucas: values,
    tribonacci: values + allpairs(),
    harmonic: values + allpairs(),
    bernoulli: values + allpairs(),
    bell: values + allpairs() + alltriples(),
    euler: values + allpairs(),
    genocchi: values + allpairs(),
    andre: values + allpairs(),
    partition: values,
    root: pairs2(0) + [(v1,0,v2) for v1,v2 in allpairs()],
    real_root: allpairs(),
    Rem: allpairs(),
    re: values,
    im: values,
    conjugate: values,
    arg: values,
    LambertW: values + allpairs(),
    erfi: values,
    erfinv: values,
    erfcinv: values,
    erf2inv: values + allpairs(),
    Ei: values,
    expint: allpairs(),
    E1: values,
    li: values,
    Li: values,
    Si: values,
    Ci: values,
    Shi: values,
    Chi: values,
    lowergamma: values + allpairs(),
    uppergamma: values + allpairs(),
    polygamma: [(0,0)] + allpairs(),
    dirichlet_eta: values + allpairs(),
    zeta: values + allpairs(),
    polylog: allpairs(),
    stieltjes: values + allpairs(),
    riemann_xi: values,
    hankel1: allpairs(),
    hankel2: allpairs(),
    jn: allpairs(),
    yn: allpairs(),
    hn1: allpairs(),
    hn2: allpairs(),
    airyai: values,
    airybi: values,
    airyaiprime: values,
    airybiprime: values,
    carmichael: values + allpairs() + alltriples(),
    marcumq: alltriples(),
    legendre: allpairs(),
    assoc_legendre: alltriples(),
    hermite: allpairs(),
    hermite_prob: allpairs(),
    chebyshevt: allpairs(),
    chebyshevu: allpairs(),
    assoc_laguerre: alltriples(),
    gegenbauer: alltriples(),
    elliptic_k: values,
    elliptic_f: allpairs(),
    elliptic_e: values + allpairs(),
    elliptic_pi: allpairs() + alltriples(),
    mathieus: alltriples(),
    mathieuc: alltriples(),
    mathieusprime: alltriples(),
    mathieucprime: alltriples(),
    # Zero should be skipped but these also fail with NameError:
    digamma: [0] + values,
    trigamma: [0] + values,
    # Fails in mpmath as well:
    multigamma: allpairs(),
    lerchphi: values + allpairs() + alltriples(),
    # ZeroDivisionError (probably due to rewrites)
    csc: [0],
    cot: [0],
    asec: [0],
    acsc: [0],
    acot: [0],
    log: pairs2(1),
    coth: [0],
    csch: [0],
    acoth: [0],
    asech: [0] + [v for v in values if v > 1],
    acsch: [0],
    # gamma pole:
    gamma: [0],
    # TypeError: laguerre() missing 1 required positional argument: 'z'
    # fails in mpmath...
    laguerre: allpairs(),
}

def nargs(f):
    if f in [sqrt, cbrt, E1]:
        return {1}
    elif f is root:
        return {2,3}
    elif f is real_root:
        return {1,2}
    elif f is loggamma:
        # KeyError: 'loggamma' (during lambdify)
        return set()
    else:
        return f.nargs


skipped = 0
passed = 0
failed = 0


def update_count(fstr, val1, val2, eps=1e-13):
    global skipped, passed, failed
    diff = complex(val1) - complex(val2)
    if abs(diff) < eps:
        passed += 1
    else:
        print(fstr, '->', diff, val1, val2)
        failed += 1


for func in functions:
    fail_values = failures.get(func, ())

    if 1 in nargs(func):
        f_mpmath = lambdify(x, func(x), modules='mpmath')
        f_arb = lambdify(x, func(x), modules='arb')
        for value in values:
            fstr = f'{func}({value})'
            #print(fstr)
            if value not in fail_values:
                mpval = f_mpmath(value)
                arbval = f_arb(arb(value))
                update_count(fstr, mpval, arbval)
            else:
                #print('skipping', fstr)
                skipped += 1

    if 2 in nargs(func):
        f_mpmath = lambdify((x, y), func(x, y), modules='mpmath')
        f_arb = lambdify((x, y), func(x, y), modules='arb')
        for value1 in values:
            for value2 in values:
                fstr = f'{func}({value1}, {value2})'
                #print(fstr)
                if (value1, value2) not in fail_values:
                    mpval = f_mpmath(value1, value2)
                    arbval = f_arb(arb(value1), arb(value2))
                    update_count(fstr, mpval, arbval)
                else:
                    #print('skipping', fstr)
                    skipped += 1

    if 3 in nargs(func):
        f_mpmath = lambdify((x, y, z), func(x, y, z), modules='mpmath')
        f_arb = lambdify((x, y, z), func(x, y, z), modules='arb')
        for value1 in values:
            for value2 in values:
                for value3 in values:
                    fstr = f'{func}({value1}, {value2}, {value3})'
                    #print(fstr)
                    if (value1, value2, value3) not in fail_values:
                        mpval = f_mpmath(value1, value2, value3)
                        arbval = f_arb(arb(value1), arb(value2), arb(value3))
                        update_count(fstr, mpval, arbval)
                    else:
                        #print('skipping', fstr)
                        skipped += 1

    if nargs(func) - {1, 2, 3} and func not in [
            carmichael, bell, Min, Max, lerchphi, # <-- any number of args
            jacobi, Ynm, Znm, betainc, betainc_regularized # <-- these take 4 args
            ]:
        assert False


print()
print('total:', skipped + passed + failed)
print('skipped:', skipped)
print('passed:', passed)
print('failed:', failed)
```

**Comment by oscarbenjamin** [comment]
We should also provide a way to use acb for complex functions.

**Comment by bjodah** [comment]
Wow, that's a very useful test suite you wrote there. I agree, we should fix these before merging this PR.

I had missed `acb`, interesting.

I'll change this back to draft status. If anyone wants to take over this PR, feel free (this was mostly an experiment on my part, and right now I don't have the spare cycles to finish this in the immediate future).

**Comment by oscarbenjamin** [comment]
Okay, I see what the Bessel function problem is. The order of arguments should be reversed:
```python
In [7]: mpmath.besselj(1, 5)
Out[7]: mpf('-0.32757913759146522')

In [8]: flint.arb(5).bessel_j(1)
Out[8]: [-0.327579137591465 +/- 4.70e-16]
```

**Comment by oscarbenjamin** [comment]
I've updated the test script above. It now reports zero failures meaning that the list of exceptions is accurate (not that everything is successful).

Note that it swaps the arguments when lambdifying bessel functions meaning that the PR currently has these arguments the wrong way round.

The output is now:
```console
$ python tarb.py 

total: 43420
skipped: 36599
passed: 6821
failed: 0
```
A lot of those skips should not really be skipped though.

File: sympy/printing/pycode.py [hunk 3]
```diff
@@ -748,3 +748,110 @@ def _print_Function(self, expr):
...
+    _kc = {k: 'arb.'+v for k, v in _known_constants_arb.items()}
+
+    def _print_Integer(self, arg):
+        return "arb(%d)" % arg
+
+    def _print_Float(self, arg):
+        return 'arb("%s")' % str(arg)
+
+    def _bessel_x(self, x, b):
+        return "{}.bessel_{}({})".format(self._print(b.args[1]), x, self._print(b.args[0]))
```
Comment (oscarbenjamin) [inline sympy/printing/pycode.py:799]: This same sort of fix is actually needed for most arb functions. Basically everything is a method and so it should be printed as a method call.

Comment (bjodah) [reply sympy/printing/pycode.py:799]: yes, that sounds reasonable. Right now I don't have time to extend this beyond the things I happen to need myself.

Comment (bjodah) [reply sympy/printing/pycode.py:799]: feel free to push directly into this branch if you feel like doing so!

**Comment by oscarbenjamin** [comment]
> At least for me it's much easier to find the time to add a few methods (tests) at a time rather, than getting full functionality all at once...

I'd be a lot happier with this approach in general for `lambdify` if the printers would reject things that they don't understand by default rather than giving confusing errors e.g.:
```python
In [7]: f = lambdify(x, airyai(x), 'numpy')

In [8]: f(1)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[8], line 1
----> 1 f(1)

File <lambdifygenerated-4>:2, in _lambdifygenerated(x)
      1 def _lambdifygenerated(x):
----> 2     return airyai(x)

NameError: name 'airyai' is not defined
```
Lambdify is already complicated for beginners to use and I constantly see people who are confused while faced with this error message.

I also just think that the code printer approach of just optimistically calling functions that have the same name is a bad idea in general. We have no idea if those functions do the same thing if it has not been checked and isn't tested e.g. (in sympy 1.11.1 - this has been fixed since):
```python
In [1]: bernoulli(10)
Out[1]: 5/66

In [2]: lambdify(x, bernoulli(x), "scipy")(10)
Out[2]: 
array([ 1.        , -0.5       ,  0.16666667,  0.        , -0.03333333,
        0.        ,  0.02380952,  0.        , -0.03333333,  0.        ,
        0.07575758])
```

Is there a simple way to make it so that lambdify rejects anything that has not been explicitly added to the printer and errors out with a good message? Something like:
```
>>> f = lambdify(x, erf(x), "arb")
...
LambdifyNotImplemented: Cannot generate arb code for the erf function.

This could be because arb does not have the erf function or because the
handler for erf has not yet been added to lambdify's ArbPrinter.
```

**Comment by bjodah** [comment]
I see your point, and I agree that your proposed behavior would be much preferable. I'll do some thinking and see if I can come up with something along those lines.

**Comment by bjodah** [comment]
Alright, this was not super easy. But I have something that might work in [d546164](https://github.com/sympy/sympy/pull/24433/commits/d5461646f9d01f4c95820667dc351eb36689a116)

The trade off is of course added complexity...

**Comment by bjodah** [comment]
This became rather complex. I'm open to creating a PR which only introduces the printers (and no functionality to lambdify) if there is interest.

**Pull request closed without merging**